move isAutoTimeAndDate / isAutoTimeZone from react-native-device-info

mikehardy opened the pull request:
<!-- Thanks for submitting a pull request! We appreciate you spending the time to work on these changes. Please follow the template so that the reviewers can easily understand what the code changes affect -->

# Summary

Howdy @zoontek  :-) - this module (which I use and which is fantastic by the way) already has time and date information in it and does it well

react-native-device-info also had some duplicate time and date functions, plus these 2 which were unique

I'd like to have one place in the react-native-community for this type of information and this seems like a natural home

So I'm purging them out of react-native-device-info in preference to here

There was no iOS implementation of them in react-native-device-info (maybe it is not possible? unknown) so this is the complete implementation, such as it is

I did not add it to the changelog as requested in this template because I could not see a changelog?

## Test Plan

<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes UI. -->

I added it to the Sync and Async example. Seems to work? Although getting the example to work with local code changes required me to run `yarn add "file:../"` before it would pick up changes...

### What's required for testing (prerequisites)?

### What are the steps to reproduce (after prerequisites)?

## Compatibility

| OS      | Implemented |
| ------- | :---------: |
| iOS     |    ❌     |
| Android |    ✅     |

## Checklist

<!-- Check completed item, when applicable, via: [X] -->

- [x] I have tested this on a device and a simulator
- [x] I added the documentation in `README.md`
- [ ] I mentioned this change in `CHANGELOG.md`
- [x] I updated the typed files (TS and Flow)
- [x] I added a sample use of the API in the example project (`example/App.js`)

zoontek commented on README.md:
```diff
@@ -267,6 +267,42 @@ console.log(RNLocalize.uses24HourClock());
 
 ---
 
+### isAutoDateAndTime()
```
Could you rename the functions `usesAutoTimeAndDate` and `usesAutoTimeZone ` to keep consistency with `uses24HourClock` or `usesMetricSystem`?

mikehardy replied:
absolutely

mikehardy replied:
done

zoontek commented on README.md:
````diff
@@ -267,6 +267,42 @@ console.log(RNLocalize.uses24HourClock());
+#### Method type
+
+```ts
+type isAutoDateAndTime = () => boolean;
````
I don't know if it's a good practice, but maybe `type isAutoDateAndTime = () => boolean | undefined;` is better since it will returns `undefined` on iOS.

mikehardy replied:
this is something I just went through on react-native-device-info - we have so many platform-specific APIs, and what to return on the non-implementing platforms? I decided to simply decide on a default return value in case the value is not known (because of permissions or something) or is not implemented, and it is returned in all cases we don't positively know. For this then I would simply platform-check in the Javascript before going down to the native module, and for all non-implementing platforms I would return false (as that is my default choice for booleans). I like that more personally than return checking but I know this is a matter of style and I am fine doing it your suggested way, or with default to false. Just wanted to offer the option. Think about it then be explicit about what style you want and I will do exactly that :-)

zoontek commented on android/src/main/java/com/reactcommunity/rnlocalize/RNLocalizeModule.java:
```diff
@@ -197,6 +198,26 @@ private WritableMap getNumberFormatSettings(Locale locale) {
     return settings;
   }
 
+  private boolean isAutoDateAndTime() {
```
I find

```java
private boolean isAutoTimeZone() {
    ContentResolver resolver = getReactApplicationContext().getContentResolver();

    return (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1
        ? Settings.Global.getInt(resolver, Settings.Global.AUTO_TIME, 0)
        : Settings.System.getInt(resolver, Settings.System.AUTO_TIME, 0)) != 0;
}
```

more elegant, but it's just because I am used to immutability 😄

mikehardy replied:
sure!

mikehardy replied:
almost forgot this one - just added it to the PR feedback commit and re-pushed

zoontek commented on example/src/AsyncExample.js:
```diff
@@ -122,6 +122,14 @@ export default class AsyncExample extends React.Component {
             name="RNLocalize.uses24HourClock()"
             value={RNLocalize.uses24HourClock()}
           />
+          <Line
```
Maybe we should display the value only if the functions are available on the platform?

```tsx
{Platform.OS === "android" && <>
  <Line
    name="RNLocalize.isAutoDateAndTime()"
    value={RNLocalize.isAutoDateAndTime()}
  />
  <Line
    name="RNLocalize.isAutoTimeZone()"
    value={RNLocalize.isAutoTimeZone()}
  />
</>}
```

mikehardy replied:
related to the above philosophy thinking on platform-specific APIs and defaults etc. I'll wait till you think on that in general.

mikehardy commented:
Thanks for considering this and after you think on it I will of course make whatever changes you like so the code is comfortable over here. I'm doing a little research on iOS APIs too, maybe we can actually have an implementation for iOS who knows?

mikehardy commented:
Okay, I searched through the API docs from Apple, and couldn't find anything. I asked the original author of the API here in case they know https://github.com/react-native-community/react-native-device-info/pull/583#issuecomment-527480557 - but I think this is simply not available on iOS, so once you decide exactly how you want to specify returns for platform-specific stuff I'll shape this up for a second review. Thanks @zoontek

zoontek commented:
@mikehardy I would rather go with:

```ts
type Option<T> = T | undefined

usesAutoDateAndTime(): Option<boolean>
usesAutoTimeZone(): Option<boolean>
```

I find it less confusing 🙂

mikehardy commented:
got it! I will do exactly that. Also the author of the original PR researched and came to the same conclusion I did - there is simply no API for iOS for these, so we're not missing anything - the PR is multi-platform "complete" such as it is. I'll tag you when I've pushed and feel like this is ready. Thanks!

mikehardy commented:
Okay @zoontek I think it's ready for a look again, here are the changes I made - some of which were in existing code, all with the idea of API-similarity:

- I added the `Option<boolean>` type you asked for, **and* I updated your existing boolean-APIs to also return it (I think you'll like that? But I could be wrong)
- I altered these two APIs to be called usesXXX instead of isXXX, because your other boolean-APIs started with uses, not is (Again, I think you'll like that? But I could be wrong)

Let me know what you think!

zoontek commented:
Looks pretty good except that `uses24HourClock` and `usesMetricSystem` returns a `boolean` 100% of the time on all platforms, so no need to wrap them in a `Option` 😅

mikehardy commented:
@zoontek  I just pushed that up with the original two methods untouched, and just my proposed additions as the 'Option<boolean>' :-)

zoontek commented:
Perfect, I merge it :)

mikehardy commented:
Wait! I always re-re-read things (can't help it) and I noticed I still had index.js with the Option wrapping in the function definitions, sorry! easy fix at least

zoontek commented:
@mikehardy I will cleanup that before the release, no problem

zoontek commented:
@mikehardy It's released: https://github.com/react-native-community/react-native-localize/releases/tag/1.2.0 :)

mikehardy commented:
Sweet! Thank you so much @zoontek - react-native-device-info v3 will be out shortly then I think

mashad6 commented:
Is it gonna be on IOS any time soon?

mikehardy commented:
@mashad6 just as soon as someone figures out 1) if it is possible then 2) proposes a PR to implement it I think. Could be as soon as today if you put in the effort and do it, in my experience @zoontek reviews + merges + releases things really fast when people put in the effort to contribute
